programs/stake/src/stake_state.rs

```diff
@@ -1023,6 +1057,34 @@ impl MergeKind {
     }
 }
 
+/// Calculate the effective credits observed for two stakes whenn merging
+///
+/// When merging two activating or active stakes, the credits observed of the
+/// merged stake is the weighted average of the two stakes' credits observed.
+fn stake_weighted_credits_observed(
+    stake: &Stake,
+    merge_lamports: u64,
+    merge_credits_observed: u64,
+) -> Option<u64> {
+    if stake.credits_observed == merge_credits_observed {
+        Some(stake.credits_observed)
+    } else {
+        let total_stake = stake.delegation.stake.checked_add(merge_lamports)? as u128;
+        let stake_weighted_credits =
+            (stake.credits_observed as u128).checked_mul(stake.delegation.stake as u128)?;
+        let merge_weighted_credits =
+            (merge_credits_observed as u128).checked_mul(merge_lamports as u128)?;
+        let total_weighted_credits = stake_weighted_credits.checked_add(merge_weighted_credits)?;
+        let effective_credits_observed = total_weighted_credits.checked_div(total_stake)?;
+        if total_weighted_credits.checked_rem(total_stake)? > 0 {
```

Comment:
... but still do we have to ceil if  `total_stake` isn't divisible by half?

this reminds of @behzadnouri 's concise math technique (https://github.com/solana-labs/solana/pull/16457#discussion_r612510176). but we're doing financial math here instead of engineering math.

so better of being verbose and rigorous, instead of succinctness?

Reply:
In this case, we're not trying to round up or down, we're *always* trying to ceiling the value if it's at all above.  This is all to make sure that the resulting stake does not get too much credit.  If we ever round down, then your merged stake gets one more credit than it truly deserves.

Simple example: if I merge a stake of 100 with credits_observed = 1001 into a stake of 200 with credits_observed = 1000, with rounding that would produce a stake of 300 with credits_observed = 1000, which would get you *more* rewards than if the stakes had stayed separated.  Does that make sense?

> this reminds of @behzadnouri 's concise math technique (#16457 (comment)). but we're doing financial math here instead of engineering math.

@ryoqun are you suggesting to add a comment that explains the math with concise variable names?

Reply:
@joncinque I think @t-nelson 's suggestion always does ceiling. so i think we're at the same page of ceiling.


> > this reminds of @behzadnouri 's concise math technique (#16457 (comment)). but we're doing financial math here instead of engineering math.
> 
> @ryoqun are you suggesting to add a comment that explains the math with concise variable names?

nope. it's just my random thought... xD no need to explain this ceiling part as a comment

Reply:
Ah I see, yes, to do the ceiling in one go, we'd have to add `total_stake - 1` before dividing by `total_stake`.  I'm fine with either

Reply:
Right, my main goal was to eliminate the second division and branching logic (Lazy Trent sees a branch and would rather eliminate it than write a test)

Reply:
good point, branches suck, will remove :smile: